File: src/py/TCPConnectionServer.py
import socket
import threading

from TCPConnectionClient import TCPConnectionClient


class TCPConnectionServer:
    """
    A server accepting connections from remote clients.
    """

    @property
    def running(self): return self.__running

    ############################################################################

    def stop(self):
        """
        Stop the server.
        """

        self.__running = False
        self.__thread.join()

    def connected(self, fn):
        """
        Register a function to be called whenever a client connects.
        """

        self.__connected.append(fn)

    def client(self):
        """
        Expect a single client to connect and return it.
        """

        raise NotImplementedError()

    ############################################################################

    def __init__(self, port: int):
        self.__so = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.__so.settimeout(1)
        self.__so.bind(("127.0.0.1", port))
        self.__so.listen(1)

        self.__connected = []

        self.__running = True

        def fn():
            while self.__running:
                try:
                    conn, addr = self.__so.accept()
                    client = TCPConnectionClient(conn, False)

                    [fn(client) for fn in self.__connected]
                except socket.timeout:
                    pass

        self.__thread = threading.Thread(target=fn)
        self.__thread.start()

File: src/kt/TCPConnectionServer.kt
import java.net.ServerSocket
import java.net.SocketTimeoutException
import java.util.concurrent.CyclicBarrier

/** A server accepting connections from remote clients. */
class TCPConnectionServer(port: Int) {
    /** Whether the server is currently active. */
    var running = true
        private set

    /** Stop the server. */
    fun stop() {
        running = false
        stopBarrier.await()
    }

    /** Register a function to be called whenever a client connects. */
    fun connected(fn: (TCPConnectionClient) -> Unit) {
        synchronized(connected) { connected.add(fn) }
    }

    /** Expect a single client to connect and return it. */
    fun client(): TCPConnectionClient {
        val barrier = CyclicBarrier(2)
        lateinit var client: TCPConnectionClient

        synchronized(awaiting) {
            awaiting.add {
                client = it
                barrier.await()
            }
        }

        barrier.await()

        return client
    }

    ////////////////////////////////////////////////////////////////////////////

    init {
        Thread {
            val server = ServerSocket(port)
            server.soTimeout = 1000

            while (running) {
                try {
                    val socket = server.accept()
                    val fn = synchronized(awaiting) {
                        awaiting.firstOrNull()?.also { awaiting.removeAt(0) } ?:
                        { client -> synchronized(connected) { connected.toList() }
                                .forEach { it(client) } }
                    }

                    fn(TCPConnectionClient(socket, false))
                }
                catch (e: SocketTimeoutException) { /* do nothing */ }
            }

            stopBarrier.await()
        } .start()
    }

    ////////////////////////////////////////////////////////////////////////////

    private val stopBarrier = CyclicBarrier(2)
    private val connected: MutableSet<(TCPConnectionClient) -> Unit> = mutableSetOf()
    private val awaiting: MutableList<(TCPConnectionClient) -> Unit> = mutableListOf()
}
